File: scripts/pipeline-v2/normalize-mof.ts
/**
 * MOF予算書・決算書CSV（download-mof-archive.tsが取得したraw ZIP）を、
 * source-preservingな行レベルレコードへ変換する（正規化のみ。集約・イベント化はderived層）。
 *
 * 仕様: 20260920_Pipeline_V2_MOF_RS統合_publicまで_最終仕様.md
 * 参照実装: Python版 pipeline_v2/normalize_mof.py に合わせている（帳票種別・列名解決・
 * 識別キーの作り方を含めて同じロジック）。
 *
 * 帳票ID→(accountType, phase)の対応:
 *   11=一般会計当初 12=特別会計当初 13=政府関係機関当初                 → initial
 *   21=一般会計補正 22=特別会計補正                                    → supplement
 *   31=一般会計概算 32=特別会計概算 33=政府関係機関概算（未使用/将来用） → provisional
 *   76=政府関係機関決算 77=一般会計決算 78=特別会計決算                 → settlement
 *
 * 提出版/成立版の区別はディレクトリ名（`{year}_teishutsu`か否か）で判定する
 * （bb.mof.go.jpのarchiveページのURL構造そのまま。download-mof-archive.ts参照）。
 *
 * 入力: data/download/mof.go.jp/archive/{year}/{yearDir}/csv/DL*.zip
 * 出力: data/normalized/mof/fy{year}/{budget-items.jsonl,manifest.json}
 *
 * 使い方: npx tsx scripts/pipeline-v2/normalize-mof.ts [year...]
 *   （年度省略時は 2024 2025）
 */
import * as fs from 'fs';
import * as path from 'path';
import { listZipEntries, readZipEntryText } from '@/scripts/zip-reader';
import { parseCsv as parseQuoteAwareCsv } from './lib/csv';
import { stableId, normalizeText } from './lib/stable-id';
import { parseIntValue, yenFromThousand } from './lib/parse';
import { writeJsonl, writeJson } from './lib/jsonl';
import { scopeOf, sectionNaturalKey, legacySectionKey, scopeNameItemKey, findHeader, standardAmountColumn, isExpenditureHeaders } from './lib/mof-keys';
import type { MofBudgetItemRecord, MofAccountType, MofPhase, MofBudgetStatus, SourceRef } from './types';

const ZIP_RE = /^DL(\d{4})(\d{2})(\d{3})\.zip$/;

const DOCUMENT_KINDS: Record<string, { accountType: MofAccountType; phase: MofPhase }> = {
  '11': { accountType: 'general', phase: 'initial' },
  '12': { accountType: 'special', phase: 'initial' },
  '13': { accountType: 'agency', phase: 'initial' },
  '21': { accountType: 'general', phase: 'supplement' },
  '22': { accountType: 'special', phase: 'supplement' },
  '31': { accountType: 'general', phase: 'provisional' },
  '32': { accountType: 'special', phase: 'provisional' },
  '33': { accountType: 'agency', phase: 'provisional' },
  '76': { accountType: 'agency', phase: 'settlement' },
  '77': { accountType: 'general', phase: 'settlement' },
  '78': { accountType: 'special', phase: 'settlement' },
};

interface MofDocument {
  fiscalYear: number;
  kind: string;
  sequence: number;
  revision: number | null;
  accountType: MofAccountType;
  phase: MofPhase;
  budgetStatus: MofBudgetStatus;
  releaseDir: string;
  zipPath: string;
}

/** raw_root相対パス（source refに使う。Python版のrelpathと同じ発想） */
function relFromRawRoot(rawRoot: string, absPath: string): string {
  return path.relative(rawRoot, absPath).split(path.sep).join('/');
}

function discoverMofDocuments(rawRoot: string, years?: Set<number>): MofDocument[] {
  const archiveRoot = path.join(rawRoot, 'mof.go.jp', 'archive');
  if (!fs.existsSync(archiveRoot)) return [];
  const docs: MofDocument[] = [];
  for (const yearDirName of fs.readdirSync(archiveRoot)) {
    const yearPath = path.join(archiveRoot, yearDirName);
    if (!fs.statSync(yearPath).isDirectory()) continue;
    for (const releaseDir of fs.readdirSync(yearPath)) {
      const csvDir = path.join(yearPath, releaseDir, 'csv');
      if (!fs.existsSync(csvDir)) continue;
      for (const zipName of fs.readdirSync(csvDir)) {
        const m = ZIP_RE.exec(zipName);
        if (!m) continue;
        const fiscalYear = Number(m[1]);
        if (years && !years.has(fiscalYear)) continue;
        const kind = m[2];
        const doc = DOCUMENT_KINDS[kind];
        if (!doc) continue;
        const sequence = Number(m[3]);
        const submitted = releaseDir.endsWith('_teishutsu');
        let budgetStatus: MofBudgetStatus;
        if (doc.phase === 'initial' || doc.phase === 'provisional') budgetStatus = submitted ? 'submitted' : 'enacted';
        else if (doc.phase === 'settlement') budgetStatus = 'settled';
        else budgetStatus = submitted ? 'submitted' : 'published';
        docs.push({
          fiscalYear,
          kind,
          sequence,
          revision: doc.phase === 'supplement' ? sequence : null,
          accountType: doc.accountType,
          phase: doc.phase,
          budgetStatus,
          releaseDir,
          zipPath: path.join(csvDir, zipName),
        });
      }
    }
  }
  return docs;
}

/** ZIP内の歳出表エントリを取得する */
function expenditureZipEntry(zipPath: string): { entry: string; rows: Record<string, string>[] } {
  for (const entry of listZipEntries(zipPath).filter(e => e.toLowerCase().endsWith('.csv'))) {
    const rows = parseCsv(readZipEntryText(zipPath, entry));
    if (rows.length > 0 && isExpenditureHeaders(Object.keys(rows[0]))) return { entry, rows };
  }
  throw new Error(`歳出表CSVが見つかりません: ${zipPath}`);
}

/**
 * quote対応のlib/csv.tsを使う（従来の素朴なsplit(',')は、値がクォート付きカンマや
 * 改行を含む場合に列がずれる可能性があった。2026-09-20 CodeRabbit指摘）。
 * ヘッダー・値をtrimし、空ヘッダー列を除外する従来の挙動は維持する
 */
function parseCsv(content: string): Record<string, string>[] {
  return parseQuoteAwareCsv(content).map(row => {
    const trimmed: Record<string, string> = {};
    for (const [key, value] of Object.entries(row)) {
      const header = key.trim();
      if (header) trimmed[header] = value.trim();
    }
    return trimmed;
  });
}

function normalizeMofDocument(rawRoot: string, doc: MofDocument): MofBudgetItemRecord[] {
  const { entry, rows } = expenditureZipEntry(doc.zipPath);
  if (rows.length === 0) return [];
  const headers = Object.keys(rows[0]);
  const { phase, accountType } = doc;

  const standardCol = phase === 'initial' || phase === 'provisional' ? standardAmountColumn(headers) : undefined;
  const previousCol = findHeader(headers, ['前年度予算額']);
  const compareCol = findHeader(headers, ['比較増△減額']);

  const baseCol = phase === 'supplement' ? findHeader(headers, ['成立予算額']) : undefined;
  const addCol = phase === 'supplement' ? findHeader(headers, ['補正', '追加額']) : undefined;
  const reductionCol = phase === 'supplement' ? findHeader(headers, ['補正', '修正減少額']) : undefined;
  const deltaCol = phase === 'supplement' ? findHeader(headers, ['補正', '差引額']) : undefined;
  const revisedCol = phase === 'supplement' ? headers.find(h => h.startsWith('改') && (h.includes('予算額') || h.includes('予定額'))) : undefined;

  const relZipPath = relFromRawRoot(rawRoot, doc.zipPath);
  const out: MofBudgetItemRecord[] = [];
  rows.forEach((row, i) => {
    const rowNumber = i + 2; // ヘッダー行の次から2行目起算（Python版に合わせる）
    const scope = scopeOf(row, accountType);
    const sectionCode = (row['項コード'] ?? '').trim();
    const sectionName = (row['項名'] ?? '').trim();
    const subItemName = (row['目名'] ?? '').trim();
    const subItemCode = (row['目番号'] ?? row['目コード'] ?? row['目別分類コード'] ?? '').trim();
    const sectionKey = sectionNaturalKey(accountType, scope, sectionCode, sectionName);
    const legacyKey = legacySectionKey(accountType, scope, sectionCode);
    const scopeNameKey = scopeNameItemKey(accountType, scope, sectionName, subItemName);
    const itemKey = [sectionKey, normalizeText(subItemCode), normalizeText(subItemName)].join('|');

    const source: SourceRef = {
      domain: 'mof.go.jp',
      path: relZipPath,
      file: path.basename(doc.zipPath),
      zipEntry: entry,
      rowNumber,
    };

    const rec: MofBudgetItemRecord = {
      schemaVersion: 2,
      recordType: 'mof_budget_item',
      recordId: stableId([relZipPath, entry, rowNumber], 'mofrow_'),
      fiscalYear: doc.fiscalYear,
      phase,
      budgetStatus: doc.budgetStatus,
      revision: doc.revision,
      accountType,
      ...scope,
      sectionCode,
      sectionName,
      subItemCode,
      subItemName,
      sectionNaturalKey: sectionKey,
      legacySectionKey: legacyKey,
      itemNaturalKey: itemKey,
      scopeNameItemKey: scopeNameKey,
      source,
    };

    if (phase === 'initial' || phase === 'provisional') {
      rec.amountYen = yenFromThousand(row[standardCol!]);
      rec.previousAmountYen = yenFromThousand(previousCol ? row[previousCol] : undefined);
      rec.differenceYen = yenFromThousand(compareCol ? row[compareCol] : undefined);
      rec.sourceAmountColumn = standardCol ?? null;
    } else if (phase === 'supplement') {
      rec.baseAmountYen = yenFromThousand(baseCol ? row[baseCol] : undefined);
      rec.supplementAdditionYen = yenFromThousand(addCol ? row[addCol] : undefined);
      rec.supplementReductionYen = yenFromThousand(reductionCol ? row[reductionCol] : undefined);
      rec.supplementDeltaYen = yenFromThousand(deltaCol ? row[deltaCol] : undefined);
      rec.revisedAmountYen = yenFromThousand(revisedCol ? row[revisedCol] : undefined);
      rec.sourceAmountColumns = {
        base: baseCol ?? null,
        addition: addCol ?? null,
        reduction: reductionCol ?? null,
        delta: deltaCol ?? null,
        revised: revisedCol ?? null,
      };
    } else if (phase === 'settlement') {
      if (accountType === 'agency') {
        rec.budgetAmountYen = parseIntValue(row['支出予算額(円)']);
        rec.carryoverInYen = parseIntValue(row['前年度繰越額(円)']);
        rec.reserveUseYen = parseIntValue(row['予備費使用額(円)']);
        rec.budgetRuleIncreaseYen = parseIntValue(row['予算総則の規定による経費増額(円)']);
        rec.reallocationYen = parseIntValue(row['流用等増△減額(円)']);
        rec.transferAdjustmentYen = 0;
        rec.currentBudgetYen = parseIntValue(row['支出予算現額(円)']);
        rec.spentYen = parseIntValue(row['支出済額(円)']);
        rec.carryoverOutYen = parseIntValue(row['翌年度繰越額(円)']);
        rec.unusedYen = parseIntValue(row['不用額(円)']);
      } else {
        rec.budgetAmountYen = parseIntValue(row['歳出予算額(円)']);
        rec.carryoverInYen = parseIntValue(row['前年度繰越額(円)']);
        rec.reserveUseYen = parseIntValue(row['予備費使用額(円)']);
        rec.budgetRuleIncreaseYen = parseIntValue(row['予算総則の規定による経費増額(円)']);
        rec.reallocationYen = parseIntValue(row['流用等増△減額(円)']);
        rec.transferAdjustmentYen = parseIntValue(row['予算決定後移替増△減額(円)']);
        rec.currentBudgetYen = parseIntValue(row['歳出予算現額(円)']);
        rec.spentYen = parseIntValue(row['支出済歳出額(円)']);
        rec.carryoverOutYen = parseIntValue(row['翌年度繰越額(円)']);
        rec.unusedYen = parseIntValue(row['不用額(円)']);
      }
    }
    out.push(rec);
  });
  return out;
}

function sortKey(r: MofBudgetItemRecord): string {
  return [r.phase, r.budgetStatus, r.revision ?? 0, r.accountType, r.sectionNaturalKey, r.subItemName, r.recordId].join('\x1f');
}

function processYear(rawRoot: string, outputRoot: string, year: number, docs: MofDocument[]): void {
  console.log(`\n=== MOF normalize: fiscalYear=${year} ===`);
  const rows: MofBudgetItemRecord[] = [];
  const inputSummaries: { phase: MofPhase; budgetStatus: MofBudgetStatus; revision: number | null; accountType: MofAccountType; path: string; rowCount: number }[] = [];
  for (const doc of docs) {
    const docRows = normalizeMofDocument(rawRoot, doc);
    rows.push(...docRows);
    inputSummaries.push({
      phase: doc.phase, budgetStatus: doc.budgetStatus, revision: doc.revision, accountType: doc.accountType,
      path: relFromRawRoot(rawRoot, doc.zipPath), rowCount: docRows.length,
    });
    console.log(`  [${path.basename(doc.zipPath)} / ${doc.releaseDir}] ${doc.phase}/${doc.budgetStatus} ${docRows.length}件`);
  }
  rows.sort((a, b) => (sortKey(a) < sortKey(b) ? -1 : sortKey(a) > sortKey(b) ? 1 : 0));

  const yearDir = path.join(outputRoot, 'normalized', 'mof', `fy${year}`);
  const rowCount = writeJsonl(path.join(yearDir, 'budget-items.jsonl'), rows);

  const phaseCounts = new Map<string, number>();
  for (const r of rows) {
    const key = `${r.phase}\x1f${r.budgetStatus}\x1f${r.accountType}`;
    phaseCounts.set(key, (phaseCounts.get(key) ?? 0) + 1);
  }
  const phaseCountList = [...phaseCounts.entries()].map(([key, count]) => {
    const [phase, budgetStatus, accountType] = key.split('\x1f');
    return { phase, budgetStatus, accountType, count };
  });

  writeJson(path.join(yearDir, 'manifest.json'), {
    schemaVersion: 2,
    fiscalYear: year,
    rowCount,
    phaseCounts: phaseCountList,
    inputs: inputSummaries,
  });
  console.log(`  budget-items.jsonl: ${rowCount}行`);
}

function main(): void {
  const years = process.argv.slice(2).map(Number).filter(n => !Number.isNaN(n));
  const targetYears = years.length > 0 ? new Set(years) : new Set([2024, 2025]);

  const rawRoot = path.join('data', 'download');
  const outputRoot = 'data';
  const docs = discoverMofDocuments(rawRoot, targetYears);
  const byYear = new Map<number, MofDocument[]>();
  for (const doc of docs) {
    const list = byYear.get(doc.fiscalYear) ?? [];
    list.push(doc);
    byYear.set(doc.fiscalYear, list);
  }
  for (const year of [...targetYears].sort()) {
    processYear(rawRoot, outputRoot, year, byYear.get(year) ?? []);
  }
}

main();

File: docs/tasks/20260919_0900_marumie-rssystem × ChatGPT 分析プラクティス/pipeline-v2-reference/pipeline_v2/normalize_mof.py
from __future__ import annotations

import re
from collections import Counter, defaultdict
from pathlib import Path
from typing import Any

from .common import (
    SCHEMA_VERSION,
    expenditure_zip_entry,
    find_header,
    normalize_text,
    parse_int,
    relpath,
    source_ref,
    stable_id,
    write_json,
    write_jsonl,
    yen_from_thousand,
)

ZIP_RE = re.compile(r"^DL(?P<year>\d{4})(?P<kind>\d{2})(?P<seq>\d{3})\.zip$")

DOCUMENT_KINDS: dict[str, tuple[str, str]] = {
    "11": ("general", "initial"),
    "12": ("special", "initial"),
    "13": ("agency", "initial"),
    "21": ("general", "supplement"),
    "22": ("special", "supplement"),
    "31": ("general", "provisional"),
    "32": ("special", "provisional"),
    "33": ("agency", "provisional"),
    "76": ("agency", "settlement"),
    "77": ("general", "settlement"),
    "78": ("special", "settlement"),
}


def discover_mof_documents(raw_root: Path, years: set[int] | None = None) -> list[dict[str, Any]]:
    archive = raw_root / "mof.go.jp" / "archive"
    docs: list[dict[str, Any]] = []
    if not archive.exists():
        return docs
    for zip_path in sorted(archive.glob("*/*/csv/DL*.zip")):
        m = ZIP_RE.match(zip_path.name)
        if not m:
            continue
        fiscal_year = int(m.group("year"))
        if years and fiscal_year not in years:
            continue
        kind = m.group("kind")
        if kind not in DOCUMENT_KINDS:
            continue
        account_type, phase = DOCUMENT_KINDS[kind]
        seq = int(m.group("seq"))
        release_dir = zip_path.parent.parent.name
        submitted = release_dir.endswith("_teishutsu")
        if phase in {"initial", "provisional"}:
            status = "submitted" if submitted else "enacted"
        elif phase == "settlement":
            status = "settled"
        else:
            status = "submitted" if submitted else "published"
        docs.append(
            {
                "fiscalYear": fiscal_year,
                "kind": kind,
                "sequence": seq,
                "revision": seq if phase == "supplement" else None,
                "accountType": account_type,
                "phase": phase,
                "budgetStatus": status,
                "releaseDir": release_dir,
                "path": zip_path,
            }
        )
    return docs


def _standard_amount_column(headers: list[str], fiscal_year: int) -> str:
    # First-year amount columns vary: 予算額 / 要求額 / 予定額 and optional (千円).
    candidates = []
    for h in headers:
        if "前年度" in h or "比較" in h or "成立" in h or h.startswith("改"):
            continue
        if "年度" not in h:
            continue
        if any(word in h for word in ("予算額", "要求額", "予定額")):
            candidates.append(h)
    if not candidates:
        raise ValueError(f"standard amount column not found: {headers}")
    return candidates[0]


def _scope(row: dict[str, str], account_type: str) -> dict[str, str]:
    if account_type == "general":
        return {
            "ministry": row.get("所管", ""),
            "organization": row.get("組織", ""),
            "specialAccount": "",
            "subAccount": "",
            "agency": "",
        }
    if account_type == "special":
        return {
            "ministry": row.get("所管", ""),
            "organization": "",
            "specialAccount": row.get("特別会計", ""),
            "subAccount": row.get("勘定", ""),
            "agency": "",
        }
    return {
        "ministry": "",
        "organization": "",
        "specialAccount": "",
        "subAccount": row.get("業務", ""),
        "agency": row.get("政府関係機関", ""),
    }


def _section_natural_key(account_type: str, scope: dict[str, str], code: str, name: str) -> str:
    if account_type == "general":
        parts = [account_type, scope["ministry"], scope["organization"], code, name]
    elif account_type == "special":
        parts = [account_type, scope["ministry"], scope["specialAccount"], scope["subAccount"], code, name]
    else:
        parts = [account_type, scope["agency"], scope["subAccount"], code, name]
    return "|".join(normalize_text(p) for p in parts)


def _legacy_section_key(account_type: str, scope: dict[str, str], code: str) -> str:
    # Mirrors the code-centric identity that can collapse same-code/different-name special-account sections.
    if account_type == "general":
        parts = [account_type, scope["ministry"], scope["organization"], code]
    elif account_type == "special":
        parts = [account_type, scope["ministry"], scope["specialAccount"], scope["subAccount"], code]
    else:
        parts = [account_type, scope["agency"], scope["subAccount"], code]
    return "|".join(normalize_text(p) for p in parts)


def _scope_name_item_key(account_type: str, scope: dict[str, str], section_name: str, sub_item_name: str) -> str:
    if account_type == "general":
        parts = [account_type, scope["ministry"], scope["organization"], section_name, sub_item_name]
    elif account_type == "special":
        parts = [account_type, scope["ministry"], scope["specialAccount"], scope["subAccount"], section_name, sub_item_name]
    else:
        parts = [account_type, scope["agency"], scope["subAccount"], section_name, sub_item_name]
    return "|".join(normalize_text(p) for p in parts)


def normalize_mof_document(raw_root: Path, doc: dict[str, Any]) -> list[dict[str, Any]]:
    zip_path: Path = doc["path"]
    entry, rows = expenditure_zip_entry(zip_path)
    if not rows:
        return []
    headers = list(rows[0].keys())
    phase = doc["phase"]
    account_type = doc["accountType"]

    standard_col = _standard_amount_column(headers, doc["fiscalYear"]) if phase in {"initial", "provisional"} else None
    previous_col = find_header(headers, "前年度予算額")
    compare_col = find_header(headers, "比較増△減額")

    base_col = find_header(headers, "成立予算額") if phase == "supplement" else None
    add_col = find_header(headers, "補正", "追加額") if phase == "supplement" else None
    reduction_col = find_header(headers, "補正", "修正減少額") if phase == "supplement" else None
    delta_col = find_header(headers, "補正", "差引額") if phase == "supplement" else None
    revised_col = None
    if phase == "supplement":
        revised_col = next((h for h in headers if h.startswith("改") and ("予算額" in h or "予定額" in h)), None)

    out: list[dict[str, Any]] = []
    for row_number, row in enumerate(rows, start=2):
        scope = _scope(row, account_type)
        section_code = (row.get("項コード") or "").strip()
        section_name = (row.get("項名") or "").strip()
        sub_item_name = (row.get("目名") or "").strip()
        sub_item_code = (
            row.get("目番号")
            or row.get("目コード")
            or row.get("目別分類コード")
            or ""
        ).strip()
        section_key = _section_natural_key(account_type, scope, section_code, section_name)
        legacy_section_key = _legacy_section_key(account_type, scope, section_code)
        scope_name_key = _scope_name_item_key(account_type, scope, section_name, sub_item_name)
        item_key = "|".join([section_key, normalize_text(sub_item_code), normalize_text(sub_item_name)])

        rec: dict[str, Any] = {
            "schemaVersion": SCHEMA_VERSION,
            "recordType": "mof_budget_item",
            "recordId": stable_id(relpath(zip_path, raw_root), entry, row_number, prefix="mofrow_"),
            "fiscalYear": doc["fiscalYear"],
            "phase": phase,
            "budgetStatus": doc["budgetStatus"],
            "revision": doc["revision"],
            "accountType": account_type,
            **scope,
            "sectionCode": section_code,
            "sectionName": section_name,
            "subItemCode": sub_item_code,
            "subItemName": sub_item_name,
            "sectionNaturalKey": section_key,
            "legacySectionKey": legacy_section_key,
            "itemNaturalKey": item_key,
            "scopeNameItemKey": scope_name_key,
            "source": source_ref(raw_root, zip_path, entry, row_number),
        }

        if phase in {"initial", "provisional"}:
            rec.update(
                {
                    "amountYen": yen_from_thousand(row.get(standard_col or "")),
                    "previousAmountYen": yen_from_thousand(row.get(previous_col or "")),
                    "differenceYen": yen_from_thousand(row.get(compare_col or "")),
                    "sourceAmountColumn": standard_col,
                }
            )
        elif phase == "supplement":
            rec.update(
                {
                    "baseAmountYen": yen_from_thousand(row.get(base_col or "")),
                    "supplementAdditionYen": yen_from_thousand(row.get(add_col or "")),
                    "supplementReductionYen": yen_from_thousand(row.get(reduction_col or "")),
                    "supplementDeltaYen": yen_from_thousand(row.get(delta_col or "")),
                    "revisedAmountYen": yen_from_thousand(row.get(revised_col or "")),
                    "sourceAmountColumns": {
                        "base": base_col,
                        "addition": add_col,
                        "reduction": reduction_col,
                        "delta": delta_col,
                        "revised": revised_col,
                    },
                }
            )
        elif phase == "settlement":
            if account_type == "agency":
                rec.update(
                    {
                        "budgetAmountYen": parse_int(row.get("支出予算額(円)")),
                        "carryoverInYen": parse_int(row.get("前年度繰越額(円)")),
                        "reserveUseYen": parse_int(row.get("予備費使用額(円)")),
                        "budgetRuleIncreaseYen": parse_int(row.get("予算総則の規定による経費増額(円)")),
                        "reallocationYen": parse_int(row.get("流用等増△減額(円)")),
                        "transferAdjustmentYen": 0,
                        "currentBudgetYen": parse_int(row.get("支出予算現額(円)")),
                        "spentYen": parse_int(row.get("支出済額(円)")),
                        "carryoverOutYen": parse_int(row.get("翌年度繰越額(円)")),
                        "unusedYen": parse_int(row.get("不用額(円)")),
                    }
                )
            else:
                rec.update(
                    {
                        "budgetAmountYen": parse_int(row.get("歳出予算額(円)")),
                        "carryoverInYen": parse_int(row.get("前年度繰越額(円)")),
                        "reserveUseYen": parse_int(row.get("予備費使用額(円)")),
                        "budgetRuleIncreaseYen": parse_int(row.get("予算総則の規定による経費増額(円)")),
                        "reallocationYen": parse_int(row.get("流用等増△減額(円)")),
                        "transferAdjustmentYen": parse_int(row.get("予算決定後移替増△減額(円)")),
                        "currentBudgetYen": parse_int(row.get("歳出予算現額(円)")),
                        "spentYen": parse_int(row.get("支出済歳出額(円)")),
                        "carryoverOutYen": parse_int(row.get("翌年度繰越額(円)")),
                        "unusedYen": parse_int(row.get("不用額(円)")),
                    }
                )
        out.append(rec)
    return out


def normalize_mof(raw_root: Path, output_root: Path, years: set[int] | None = None) -> dict[str, Any]:
    docs = discover_mof_documents(raw_root, years)
    by_year: dict[int, list[dict[str, Any]]] = defaultdict(list)
    doc_summaries: list[dict[str, Any]] = []
    for doc in docs:
        rows = normalize_mof_document(raw_root, doc)
        by_year[doc["fiscalYear"]].extend(rows)
        doc_summaries.append(
            {
                "fiscalYear": doc["fiscalYear"],
                "phase": doc["phase"],
                "budgetStatus": doc["budgetStatus"],
                "revision": doc["revision"],
                "accountType": doc["accountType"],
                "path": relpath(doc["path"], raw_root),
                "rowCount": len(rows),
            }
        )

    manifests: dict[str, Any] = {"schemaVersion": SCHEMA_VERSION, "years": {}}
    for year, rows in sorted(by_year.items()):
        rows.sort(
            key=lambda r: (
                r["phase"],
                r["budgetStatus"],
                r["revision"] or 0,
                r["accountType"],
                r["sectionNaturalKey"],
                r["subItemName"],
                r["recordId"],
            )
        )
        year_dir = output_root / "normalized" / "mof" / f"fy{year}"
        count = write_jsonl(year_dir / "budget-items.jsonl", rows)
        phase_counts = Counter((r["phase"], r["budgetStatus"], r["accountType"]) for r in rows)
        manifest = {
            "schemaVersion": SCHEMA_VERSION,
            "fiscalYear": year,
            "rowCount": count,
            "phaseCounts": [
                {"phase": p, "budgetStatus": s, "accountType": a, "count": c}
                for (p, s, a), c in sorted(phase_counts.items())
            ],
            "inputs": [d for d in doc_summaries if d["fiscalYear"] == year],
        }
        write_json(year_dir / "manifest.json", manifest)
        manifests["years"][str(year)] = manifest
    write_json(output_root / "normalized" / "mof" / "manifest.json", manifests)
    return manifests
